### ISSUE TITLE
APF: shufflesharding unit tests time out (flake)

### ISSUE THREAD

**Issue by adtac** [kubernetes/kubernetes#98322]
**Which test(s) are flaking**: //staging/src/k8s.io/apiserver/pkg/util/shufflesharding

**Testgrid link**: https://testgrid.k8s.io/sig-release-1.20-blocking#bazel-test-1.20&include-filter-by-regex=shufflesharding

While there aren't any visible failures in the above testgrid link, we've observed flaky timeouts downstream.

**Reason for failure**: Test times out

**Anything else we need to know**:

We don't have full logs for failed unit tests, so it's hard to pinpoint exactly which test is timing out, but I believe the offending function is this:

https://github.com/kubernetes/kubernetes/blob/1bb010857445fc9b8b35e9e36a9fbbbdd80694b3/staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go#L201-L211

You will see that the test iterates `hashMax` times:

https://github.com/kubernetes/kubernetes/blob/1bb010857445fc9b8b35e9e36a9fbbbdd80694b3/staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go#L227

Summing up the hashMax for all four subtests, we see that the function iterates ~34 million times. While the test takes just ~9 seconds on my machine to complete, the CI is probably struggling with it (other tests are running in parallel and iterating 34 million times over a non-trivial loop is very CPU hungry). A successful execution of the unit test on CI takes ~3-4 minutes on average. The test doesn't have any of the usual suspects that can cause a timeout (like <-chan receive, i/o reads, wait.Poll(), time.Sleep(), etc.), so I suspect the large number of iterations to be the culprit.

I propose making the unit test smaller and quicker to complete. Hopefully someone who understands shufflesharding better than me can take over :)

/sig api-machinery
/cc @MikeSpreitzer @lavalamp @liggitt

**Comment by liggitt** [comment]
when run with `-race`, the runtime increases significantly:

```
go test  -race -c k8s.io/apiserver/pkg/util/shufflesharding 

time ./shufflesharding.test
PASS

real	0m34.366s
user	0m34.849s
sys	0m0.684s
```

that time is dominated by TestUniformDistribution

**Comment by adtac** [comment]
The test does a lot of slice appends. Maybe `-race` is constraining the allocator? There isn't much else in the test that might be slowed down by `-race`: no mutexes, no RNGs, no atomic operations, no channels, etc.

**Comment by lavalamp** [comment]
It looks like the buffer is pre-allocated to be the correct size though, so none of those appends should be resulting in allocations?

**Comment by MikeSpreitzer** [comment]
BTW, in another branch I tried taking the `append`s out of the test, instead calling `Deal` with a lambda that writes into `aHand` (very much like the implementation of `DealIntoHand`).  The `test -race` got a little slower!
